### PR TITLE
Windows: create daemon root with ACL

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

This PR ensures the Windows daemon creates the root with an ACL for Local System and Built-In Administrators. @NathanMcCauley @thaJeztah

### CODE REVIEW

**Comment by vieux** [comment]
SGTM

File: pkg/system/filesys_windows.go
Comment (johnstep) [inline pkg/system/filesys_windows.go]: Update the comment to reflect that it grants full access to administrators and system.

File: cmd/dockerd/daemon.go
Comment (jhowardmsft) [inline cmd/dockerd/daemon.go]: Fair. Updated in next push (imminent)

**Comment by jhowardmsft** [comment]
@johnstep Comments addressed. Rebased following minor merge conflict after the merge of the 'plug-ins coming out of experimental' PR just now.

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: @jhowardmsft Could you explain the moved line here?

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: Same question: can we keep this inside `NewDaemon`. From go api standpoint it doesn't make sense that this preparation is needed before calling `NewDaemon`.

Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: Is it possible to move pidfile/migrateKey instead?

File: cmd/dockerd/daemon.go
Comment (jhowardmsft) [inline cmd/dockerd/daemon.go]: It's to ensure the root is created first, before anything else. The current ordering creates other files prior to the daemon root, indirectly creating the root in the process.

File: cmd/dockerd/daemon.go
Comment (jhowardmsft) [inline cmd/dockerd/daemon.go]: Same answer as above.

Comment (jhowardmsft) [inline cmd/dockerd/daemon.go]: Actually, not in the current code flow. The startup path for the daemon is kind of strange and all over the place. I have to guarantee that the very first thing done is to create the root before any other file is created.

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: I understand that. Can you point me where in configuration loading we create it so the order initializing trust key  path becomes significant.

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: I think we can do better but let's leave that to some other PR.

**Review by johnstep — APPROVED** [review]
LGTM

File: cmd/dockerd/daemon.go
Comment (jhowardmsft) [inline cmd/dockerd/daemon.go]: Yes, agreed. I wanted to be as minimally invasive as possible and just fix the immediate issue.

File: cmd/dockerd/daemon.go
Comment (jhowardmsft) [inline cmd/dockerd/daemon.go]: I'd have to debug it again, can't immediately see it. But I know it was the case when creating the commit back at the weekend.

**Comment by jhowardmsft** [comment]
@crosbymichael @tiborvass @cpuguy83 In the hope that this might get in ahead of the code freeze cutoff. Appreciate you're all busy though. Or asleep ;)

**Comment by tonistiigi** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @riyazdf PTAL

File: cmd/dockerd/daemon.go
Comment (riyazdf) [inline cmd/dockerd/daemon.go]: leaving a breadcrumb since we touched relevant code to this FIXME: especially from the changes in this PR it seems that we can move this to the `TrustKey` logic above.  We can do this in a followup PR.

File: cmd/dockerd/daemon_windows.go
Comment (riyazdf) [inline cmd/dockerd/daemon_windows.go]: nit: the `\` might be redundant for `filepath.Join` in windows

File: daemon/daemon_windows.go
Comment (riyazdf) [inline daemon/daemon_windows.go]: nit: it might be worth leaving a comment about why `0` is acceptable for Windows since that argument isn't used as permissions

File: pkg/system/filesys_windows.go
Comment (riyazdf) [inline pkg/system/filesys_windows.go]: nit: might be worth leaving a comment explaining what this expression semantically means

**Review by riyazdf — APPROVED** [review]
code LGTM with nits.

However, I think we should add windows-specific tests around these permissions tests to ensure there are correct and there are no future regressions.  This can happen in a follow-up PR.

**Comment by thaJeztah** [comment]
@jhowardmsft can you do a follow up to address those nits? I'll move this to "merge"

**Comment by jhowardmsft** [comment]
Yes.

**Pull request merged**
